cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -84,32 +85,278 @@ namespace util {
+};
+
+/// Decoder class for Bit packing encoded data.
+template <typename T>
+class BitPackedDecoder {
+ public:
+  /// The type in which the data should be decoded.
+  using value_type = T;
+  /// The type of run that can be decoded.
+  using run_type = BitPackedRun;
+  using values_count_type = run_type::values_count_type;
+  using bit_size_type = run_type::bit_size_type;
+
+  BitPackedDecoder() noexcept = default;
+
+  explicit BitPackedDecoder(run_type const& run) noexcept;
+
+  void Reset(run_type const& run) noexcept;
+
+  /// Return the number of values that can be advanced.
+  [[nodiscard]] constexpr values_count_type Remaining() const;
+
+  /// Return the size in bit in which each encoded value is written.
+  [[nodiscard]] constexpr bit_size_type ValueBitWidth() const;
+
+  /// Try to advance by as many values as provided.
+  /// Return the number of values skipped.
+  [[nodiscard]] values_count_type Advance(values_count_type batch_size);
+
+  /// Get the next value and return false if there are no more.
+  [[nodiscard]] bool Get(value_type* out_value);
+
+  /// Get a batch of values return the number of decoded elements.
+  [[nodiscard]] values_count_type GetBatch(value_type* out, values_count_type batch_size);
+
+ private:
+  ::arrow::bit_util::BitReader bit_reader_ = {};
+  bit_size_type value_bit_width_ = 0;
+  values_count_type remaining_count_ = 0;
+
+  static_assert(std::is_integral_v<value_type>,
+                "This class makes assumptions about integer endianness and padding");
+};
+
+/// Decoder class for RLE encoded data.
+template <typename T>
+class RleBitPackedDecoder {
+ public:
+  /// The type in which the data should be decoded.
+  using value_type = T;
+  using byte = RleBitPackedParser::byte;
+  using raw_data_const_pointer = RleBitPackedParser::raw_data_const_pointer;
+  using raw_data_size_type = RleBitPackedParser::raw_data_size_type;
+  using bit_size_type = RleBitPackedParser::bit_size_type;
+  using dynamic_run_type = RleBitPackedParser::dynamic_run_type;
+  /// The type of the size of either run, between 1 and 2^31-1 as per Parquet spec
+  using values_count_type = int32_t;
+
+  RleBitPackedDecoder() noexcept = default;
+
+  /// Create a decoder object.
+  ///
+  /// data and data_size are the raw bytes to decode.
+  /// value_bit_width is the size in bits of each encoded value.
+  RleBitPackedDecoder(raw_data_const_pointer data, raw_data_size_type data_size,
+                      bit_size_type value_bit_width) noexcept;
+
+  void Reset(raw_data_const_pointer data, raw_data_size_type data_size,
+             bit_size_type value_bit_width_) noexcept;
+
+  /// Whether there is still runs to iterate over.
+  ///
+  /// WARN: Due to lack of proper error handling, iteration with Get methods could return
+  /// no data while the parser is not exhausted.
+  /// This is how one can check for errors.
+  [[nodiscard]] bool Exhausted() const;
```

Comment:
Nit, but we would use `snake_case` for such cheap properties.
```suggestion
  [[nodiscard]] bool exhausted() const;
```

Reply:
Just a ping on this :)